pom.xml

```diff
@@ -43,6 +43,10 @@
       <url>https://github.com/eclipse-ee4j/glassfish</url>
     </scm>
 
+    <properties>
+        <project.build.outputTimestamp>2020-12-19T17:24:00Z</project.build.outputTimestamp>
```

Comment:
We want to know when the artifact was built and I am sure it wasn't in 2020.

Reply:
> We want to know when 

- I don't. https://github.com/eclipse-ee4j/glassfish/pull/24316#issuecomment-1445037472

Reply:
it will be updated when you use `mvn versions:set`

Reply:
I opened the link from the description and I think I understand the issue. It is a timestamp, TIMESTAMP. It doesn't matter if it is in the signature file, pom or properties, Maven Central just wants that it doesn't change in the release.

I will try something on monday. 

Settting that to 2020 or any other value for snapshots is clearly a nonsense.

Reply:
you can have a look at https://maven.apache.org/guides/mini/guide-reproducible-builds.html#faq

your reaction is classical. you're discovering that a timestamp is a timestamp that does not always represent "now" from the current build (which is by definition not reproducible: "now" from 2 builds on the same code is not the same): a timestamp is a timestamp of "something"

> or any other value for snapshots is clearly a nonsense

I worked on it for sufficient time to tell you that it is NOT a nonsense, but a change from previous habits.
For some people who feel unhappy, they set this timestamp to the last Git commit: makes them less unhappy from a logical perspective.
From a build cache perspective in a context like yours of a massive multi-module build, it's not the best choice: many modules will have their output checksum changed even if no code changed.
FYI, on that question, Gradle did a straightforward choice: timestamp is a fixed value = start of Unix EPOCH...

At least, here, the timestamp is updated on release: releases give you an expected semantics on the timestamp
And you'll have to accept that the timestamp of the start of a new spring is the best option to void rebuilding content when it's not necessary

Reply:
So now the definition of the "repeatable build" is just having a constant timestamp in pom.xml, right? It doesn't matter if jar files have different content (and they do). It is not just illogical, it is simply wrong.

> your reaction is classical. you're discovering that a timestamp is a timestamp that does not always represent "now" from the current build (which is by definition not reproducible: "now" from 2 builds on the same code is not the same): a timestamp is a timestamp of "something"

If it is classical, it is probably a signal that your idea may be wrong. How setting a timestamp to a constant makes builds repeatable? It doesn't, it is just a fake date.

> many modules will have their output checksum changed even if no code changed.

Yes! Exactly - when checksums changed, we don't have 100% reproducible build. And then I want a human readable timestamp from which I can quickly see when I or someone else built the artifact; then I can guess a direction of what happened. 

Maven Central prohibits reuploading artifacts if I remember well, but that's not true for other repositories. Then is useful to find who, where, when and how someone created the duplicit release. And then we have snapshots, where you do that nearly on a daily basis.

Your change affects what we see this in logs; now there would be year 2020:
```
Running GlassFish Version: Eclipse GlassFish 7.0.3-SNAPSHOT (branch: master, commit: 1cdea232a94fadebf51523440c31afa74b29c38d, timestamp: 2023-03-30T11:16:14Z)]]
```

> Gradle did 

I don't know even how Gradle uses timestamps, because it is always a timestamp of some event. As you mentioned - it depends on the usage. In our case it is a build timestamp. And as builds are never 100% repeatable, which is proven by changing checksums of artifacts, it should change too. But ok, current hype is that build timestamp is not a timestamp.

> the timestamp is updated on release

In fact it is NOT updated. You just set the really old date here. We release a version nearly every month and the timestamp was this old date for several versions.

> it will be updated when you use mvn versions:set

1. MojoHaus versions plugin is not a standard plugin. It shouldn't be used in "perfect" projects, it is rather an additional tool. GlassFish is not a perfect project yet, so sometimes we use it.
2. Maven release plugin is the right plugin. Here it writes about our issue: https://maven.apache.org/guides/mini/guide-reproducible-builds.html#faq

-------
So let's get to something, please tell me if you agree:

1. For snapshots the value will be generated (as it is until now).
2. For a releases since 7.0.4 we will update the project and releasing process to ensure that in the released pom the value in the pom.xml is set. My current idea is this:
    1. Update the release process to use the maven release plugin as much as possible.
    2. Add your plugin to the build, so it would fail the release build if the value wouldn't be set.
    3. Update the "release howto"

I believe this would resolve the problem for both of us, because then
1. The release will be repeatable in your terminology.
2. The snapshot would change the timestamp with every build (which may contain even uncommitted changes, so using git timestamp is a mistake and using a constant means resignation).